channelflow/nse.cpp

```diff
@@ -809,39 +809,74 @@ Real viscosity(Real Reynolds, VelocityScale vscale, MeanConstraint constraint, R
     return nu;
 }
 
+// logic
+//
+// Vsuck == 0
+//   pressure constraint         =>   regular PCF/PPF formula, pressure style
+//   bulk velocity constraint    =>   regular PCF/PPF formula, bulk vel style
+// Vsuck != 0
+//   pressure constraint
+//     dPdx == 0                 =>   regular ASBL formula
+//     dPdx != 0                 =>   pressure ASBL formula
+//   bulk velocity constraint    =>   bulk velocity ASBL formula
+
 ChebyCoeff laminarProfile(Real nu, MeanConstraint constraint, Real dPdx, Real Ubulk, Real Vsuck, Real a, Real b,
                           Real ua, Real ub, int Ny) {
     ChebyCoeff u(Ny, a, b, Spectral);
 
-    if (constraint == BulkVelocity) {
-        if (abs(Vsuck) > 1e-14)
-            cferror("Combining suction with constraint BulkVelocity is not implemented yet");
-
-        u[0] = 0.125 * (ub + ua) + 0.75 * Ubulk;
-        u[1] = 0.5 * (ub - ua);
-        u[2] = 0.375 * (ub + ua) - 0.75 * Ubulk;
-
-    } else {
-        if (abs(Vsuck) < 1e-14) {
+    if (Vsuck == 0.0) {
+        if (constraint == BulkVelocity) {
+            u[0] = 0.125 * (ub + ua) + 0.75 * Ubulk;
+            u[1] = 0.5 * (ub - ua);
+            u[2] = 0.375 * (ub + ua) - 0.75 * Ubulk;
+        } else {
             dPdx *= square((b - a) / 2);
             u[0] = 0.5 * (ub + ua) - 0.25 * dPdx / nu;
             u[1] = 0.5 * (ub - ua);
             u[2] = 0.25 * dPdx / nu;
         }
+    } else {  // Vsuck != 0
+        u.setState(Physical);
+        Vector y = chebypoints(Ny, a, b);
+        Real H = b - a;
+        Real ub_ua = ub - ua;
+        Real Vsuck_nu = Vsuck / nu;
+        Real expm1_H_Vsuck_nu = expm1(-H * Vsuck_nu);  // = exp(-H*Vsuck/nu) - 1
 
-        else {
-            u.setState(Physical);
-            if (dPdx != 0)
-                cferror("Combining suction with constraint PressureGradient is not implemented yet");
-            Real delta = nu / Vsuck;
-            Vector y = chebypoints(Ny, a, b);
-            Real c1 = (ub - ua) / (exp(-b / delta) - exp(-a / delta));
-            Real c2 = ua - c1 * exp(-a / delta);
+        if (constraint == PressureGradient) {
+            Real dPdx_Vsuck = dPdx / Vsuck;
+
+            if (dPdx == 0.0) {
```

Comment:
Here as well.

Reply:
This is safe. The dPdx != 0 formula is linear in dPdx and reduces to the dPdx == 0 case when dPdx has value zero. I included both so that the classic ASBL formula appeared explicitly, but I can see it would be better to collapse them to one case and just note how one reduces to the other in a comment.

Reply:
Thanks, I see.
I guess It's better to have both formula explicitly written. It's not an expansive operation and makes the code clearer.